Add Atom CPU profiler documentation (#896)
o3de/o3de.org: content/docs/atom-guide/atom-sample-viewer/debug-menu-samples.md

---
linkTitle: Debug Menu Samples
title: Debug Menu Samples
description: Learn to use the profiling and debugging tools available for Atom Renderer with Atom Sample Viewer's debug menu samples.
toc: true
--- 

The Atom Sample Viewer debug menu samples demonstrate the various profiling and debugging tools you can use with Atom Renderer. From the debug menu, you can capture frame data, analyze threads, view stats, and visualize CPU and GPU profile data.

{{< placeholder >}}
{{< todo issue="https://github.com/o3de/o3de.org/issues/687" >}}

> {{< /todo >}}

Reviewer: The banners still exist. Are we keeping them for some reason?

Reviewer 2: I don't think we need the **Placeholder** banner. I'd keep the **To do** banner since there are more sections that needs docs.

Author: Placeholder banner removed!

> {{< placeholder >}}
> {{< todo issue="https://github.com/o3de/o3de.org/issues/687" >}}
> {{< /todo >}}

Reviewer 2: Remove the placeholder banner.

```suggestion
{{< todo issue="https://github.com/o3de/o3de.org/issues/687" >}}
{{< /todo >}}
```

### High-level Toolbar

The CPU Profiler widget has two views: a statistics view and a visualizer. These two views share a common header for controlling the profiler.

![CPU Profiler Views](/images/atom-guide/atom-sample-viewer/cpu-profiler-views.png)

| Tool | Description |
| - | - |
| **Swap to statistics/visualizer** | Swaps between the statistics and visualizer views. |
| **Resume/Pause** | Starts/stops Atom profile data collection. |
| **Capture** | Captures a single frame of CPU profile data. The collected data is saved to `/<project>/user/CpuProfiler/`. |
| **Begin/End** | Starts/stops a capture of continuous frames of CPU profile data. The collected data is saved to `/<project>/user/CpuProfiler/`. |
| **Load file** | Opens the File Picker window, which lists all of the capture files in `/<project>/user/CpuProfiler/`. The files are sorted in descending order according to modification time, with the most recent capture at the top. Refer to [Loading Captures](#loading-captures). |

### Loading Captures

![CPU Profiler File Picker](/images/atom-guide/atom-sample-viewer/cpu-profiler-file-picker.png)

> 

Reviewer: Not a big deal, extra new line here.

```suggestion
```

CPU Profiler allows you to capture CPU profiling data at a specific time-frame, which you can load and view in the statistics or visualizer views. 

To load captures of CPU profiling data:
1. Click **Load file** in the CPU Profiler's Toolbar to open the File Picker window. 
2. Select the capture that you want to load. Captures are stored as `.json` files in `/<project>/user/CpuProfiler/`.
3. Click **Load selected** to load the selected capture into memory. This displays the captured CPU profiling data in the main CPU Profiling window. 

> 

Reviewer: an extra new line

```suggestion
```

> {{< /todo >}}

Author: Done!

## CPU Profiler

> The Atom Renderer's built-in CPU Profiler widget can be used to collect running statistics and visualize profiling data collected from the renderer. 

Author: Noted on repeat usage of collect and profile, will go with Mike's variation on this one.

> To open the CPU Profiler widget, click on `Profile` &rarr; `CPU Profiler` from the ImGui debug menu.

Author: Let me add that then.

### High-level Toolbar

The CPU profiler widget has two views, a statistics view and a visualizer. These two views share a common header for controlling the profiler.

![CPU Profiler Views](/images/atom-guide/atom-sample-viewer/cpu-profiler-views.png)

> * Swap to statistics is used to swap between the statistics view and the visualizer.
> * Resume can be used to start or stop profiling data collection from Atom.
> * Capture takes a single-frame capture of CPU profiling data and saves it to a location on disk, by default into /<project>/user/CpuProfiler/.
> * Begin can be used to start a continuous capture of CPU profiling data, during which Atom is storing many frames of data. End will end the continuous capture and save the data that was collected to disk, by default into /<project>/user/CpuProfiler/.
> * Load File will bring up the file picker window which will list all of the.json files in /<project>/user/CpuProfiler/. This list of files is sorted in decreasing order according to modification time, so the most recent capture is at the top.

Reviewer: Do we want to put these in a table? I also made some slight changes in descriptions. Can we use "profile" instead of "profiling"?

```suggestion
| Tool | Description |
| - | - |
| **Swap to statistics/visualizer** | Swaps between the statistics and visualizer views. |
| **Resume/Pause** | Starts/stops Atom profile data collection. |
| **Capture** | Captures a single frame of CPU profile data. The collected data is saved to `/<project>/user/CpuProfiler/`. |
| **Begin/End** | Starts/stops a capture of continuous frames of CPU profile data. The collected data is saved to `/<project>/user/CpuProfiler/`. |
| **Load file** | Opens the File Picker window, which lists all of the capture files in `/<project>/user/CpuProfiler/`. The files are sorted in descending order according to modification time, with the most recent capture at the top. Refer to [Loading Captures](#loading-captures). |
```

### Loading Captures

![CPU Profiler File Picker](/images/atom-guide/atom-sample-viewer/cpu-profiler-file-picker.png)

* Load Selected will load the selected capture into memory, after which the data will be visible in the statistics and visualizer views.

### Statistics

![CPU Profiler Statistics](/images/atom-guide/atom-sample-viewer/cpu-profiler-statistics.png)

The statistics view focuses on processing and displaying the quantitative data collected by Atom's CPU profiler. Each row in the table represents a specific profiling region, and all columns in the table can be sorted by clicking on the column headers:

> * Group: The overall renderer area (ex. RPI, RHI, DX12) that the profiling region occurs in.
> * Region: The lower-level name of the profiling region, often the name of a function call.
> * MTPC: The mean time per call in milliseconds of the profiling region aggregated across all of the data collected while the CPU profiler widget has been active.
> * Max: The maximum execution time of the region in milliseconds during the last frame's execution. For regions that only appeared once in the last frame, this will be equal to the total time.
> * Invocations: The number of times that the profiling region was executed in the last frame.
> * Total: The overall time spent in milliseconds executing this profiling region on the last frame, summing up each invocation of the region.

Reviewer: Do we want to use a table here? Some edits to descriptions. Also, is it acceptable to use "profile" rather than "profiling" in certain circumstances as below?

```suggestion
| **Group** | The overall renderer area (RPI, RHI, or DX12, for example) of the profile region. |
} **Region** | The lower-level name of the profile region, which is often the name of a function call. |
* **MTPC** | The mean time per call in milliseconds of the profile region aggregated across all of the data collected. |
| **Max** | The maximum runtime of the region in milliseconds during the last frame's execution. For regions that only appear once in the last frame, this will be equal to the total time.
| **Invocations** | The number of times that the profile region ran in the last frame. |
| **Total** | The overall time in milliseconds this profile region required on the last frame, summing up each invocation of the region. |
```

> Rows in the table with a specific group or region name can also be searched for with the filter, which can be easily cleared by clicking  Clear Filter. Reset Table will clear all of the table's saved data and is recommended when switching samples.

Reviewer: ```suggestion
You can filter rows by entering a specific group or region name in the text field and clicking **Filter**. To remove the filter, click **Clear Filter**. To clear all of the table's saved data, which is recommended when switching samples, click **Reset Table**.
```

> Rows in the table with a specific group or region name can also be searched for with the filter, which can be easily cleared by clicking  Clear Filter. Reset Table will clear all of the table's saved data and is recommended when switching samples.

Author: Both seem great! I like Chanelle's version a bit more.

> ### Visualizer

Reviewer 2: I have two separate suggestions for this subsection, labeled as **Option A** and **Option B**.

**Option A**: Does not reorder the contents of this subsection. Minimal edit. 
**Option B**: Reformat this subsection into two texts: _conceptual_ and _action_. The conceptual paragraphs describe the visualizer and its components, such as the histogram. The action table groups all of the ways the user can interact with the visualizer to do different functions. 

These options will make more sense as you review the suggestions that follow. I believe Option B organizes the information better, so there's a clear line between "learn" and "do. Also, a user can easily refer to the actions table to recall how to do a specific function.

Author: Went with option B, the sections are now informally divided into two parts

![CPU Profiler Statistics](/images/atom-guide/atom-sample-viewer/cpu-profiler-visualizer.png)

The visualizer presents the CPU profiling data in a timeline format similar to other profilers like RAD Telemetry and Tracy. Each active render thread is laid out horizontally, with every profiling event corresponding to a block in the visualizer. Time increases going to the right.

> To scroll throughout the data, hold **RMB+Drag**. To zoom in and out, use **Control+Mouse Wheel**.

Reviewer 2: **Option A:**
I suggest reformatting this into an "**action** : description" list to make it easier to read and reference. 
```suggestion
You can navigate through the timeline using the following controls:
- **RMB+Drag**: To scroll through the data.
- **Ctrl+Mouse Wheel**: To zoom in and out.
```

Reviewer 2: **Option B**: 
Combine these interactions with the other set of interactions, such as using the **Saved Frames** slider or **Find Region** text field. (See the last **Option B** suggestion.)
```suggestion
```

Author: First one seems easier to follow

> The histogram at the top of the widget can be used to easily find frames with poor performance. The height of each individual block corresponds to the amount of time it took for the frame to execute. The orange line corresponds to a frame time of 16.6 ms, or 60 FPS, and the red line corresponds to a frame time of 33.3 ms (30 FPS).

Reviewer 2: **Option A**: 
```suggestion
The histogram at the top of the widget allows you to find frames with poor performance. The height of each individual block corresponds to the amount of time it took for the frame to run. The orange line corresponds to a frame time of 16.6 ms, or 60 FPS. The red line corresponds to a frame time of 33.3 ms, or 30 FPS.
```

Reviewer 2: **Option B:** 
Leave as is. If you follow all of **Option B** suggestions, this should be the second paragraph in this sub section.

> To quickly see the statistics for a given region, clicking on a region in the visualizer will jump to the statistics window with the targeted region's name filled in the filter. The Saved Frames slider controls the number of frames to save within the profiler. Find Region can be used as a filter to only draw specific regions of interest.

Reviewer 2: **Option A**:
```suggestion
To quickly see the statistics for a given region, click on a region in the visualizer. This switches to the statistics window and filters the data with the targeted region's name. To adjust the number of frames that you want to save within the profiler, use the **Saved Frames** slider. To filter the regions drawn in the histogram, enter a region name in the **Find Region** text field.
```

Reviewer 2: **Option B**:
 Group all the user-actions into a table like this. 

```suggestion
You can navigate the CPU profiling data in the visualizer with the following actions:

| Action | Description |
| - | - |
| **RMB+Drag** | Scrolls through the data. |
| **Ctrl+Mouse Wheel** | Zooms in and out. |
| Click on a region in the visualizer. | Switches to statistics view and filters the data by the selected region. |
| Adjust the **Saved Frames** slider. | Adjusts the number of frames that you want to save within the profiler. |
| Enter a region name in the **Find Region** text field. | Filters the regions that're drawn in the histogram. |
```

> 

Author: Done
